[GPII-3865]: Improve the way we set organization-level permissions

natarajaya opened the pull request:
Summary of changes:
* Every time `:apply_common_infra` runs, it invokes `:set_org_perms` that
  1. Removes all roles from individual user accounts (to revoke perms granted to operators by `:grant_org_admin`).
  1. For common SA, cloud-admin group and Eugene's account: it ensures that only expected set of permissions is present, removes all unexpected roles.
* Refactored `:refresh_common_infra` and related tasks.

natarajaya commented on shared/rakefiles/xk_infra.rake:
```diff
@@ -1,15 +1,16 @@
   "roles/serviceusage.serviceUsageAdmin",
-  "roles/storage.admin"
+  "roles/storage.admin",
+  "roles/viewer"
```
Where do you find this information? [Documentation](https://cloud.google.com/iam/docs/understanding-roles) says nothing about viewer role deprecation...

mrtyler replied:
I think I was conflating "primitive" with "deprecated.

However, it is a best practice to avoid using primitive roles: https://forsetisecurity.org/docs/latest/concepts/best-practices.html

However however, this use case is a reasonable exception. I'd say the second point from this doc applies: https://cloud.google.com/iam/docs/faq#when_would_i_use_primitive_roles

In conclusion, I retract this suggestion.

mrtyler commented:
LGTM unless you want to talk about unit tests :).

natarajaya commented:
@mrtyler The code was not quite ready for the review, hence `work-in-progress` label, but thanks for reviewing it anyway! :)

I updated top comment with the summary of changes.

> LGTM unless you want to talk about unit tests

Sure, I also wanted to discuss the general strategy of enforcing org-level permissions with the team, added parking lot item.

stepanstipl commented:
I would, in general, prefer to move this to Terraform. 

Apart from the overhead of writing and maintaining custom code for something that Terraform can do, there are several issues which I think would make the TF a better option, amongst the main ones:
- Lack of error handling (and retry logic)
- Lack of tests
- Very narrow scope - for example using `--filter bindings.members:user:` will result in only removing individual user permission, but what about permissions assigned to service accounts, groups, whole domain or special `allUsers`?

Otherwise looks good. Is there a reason to mix use of `%x{}` and `sh` for executing shell commands?

natarajaya commented:
> prefer to move this to Terraform

@stepanstipl We would still need to have a code that enables services, creates infra primitives and sets required permissions so Terraform can run. And `xk_infra` code is doing pretty much all things that [gcp-project-init](https://github.com/exekube/exekube/blob/master/modules/gcp-project-init/gcp-project-init) module used to do, in a bit more sophisticated manner.

> Lack of error handling (and retry logic)

Well, it handles the "normal" conditions (all possible existing resource states) pretty well, in case of networking issues, GCP issues, and other unforeseen circumstances – it fails the pipeline and requires manual action to retry.

> Lack of tests

We have a parking lot item with this topic.

> Very narrow scope - for example using --filter bindings.members:user: will result in only removing individual user permission

Which is the goal of this task – to remove permissions added by `:grant_org_admin`. Also, previous version of this code was only raising an exception when some permissions did not match with expected state of things, so it is already a progress :) Enforcing full set of org-level (and project-level) permissions in one place is a difficult topic, let's discuss it further with the rest of the team.

> Is there a reason to mix use of %x{} and sh for executing shell commands

`%x{} ` is used when we need to capture command's output.

stepanstipl commented:
Thanks @natarajaya 

> @stepanstipl We would still need to have a code that enables services, creates infra primitives and sets required permissions so Terraform can run. And xk_infra code is doing pretty much all things that gcp-project-init module used to do, in a bit more sophisticated manner.

No, we would also use terraform for this ;). (If you remember my pure Terraform concept, it's done there - https://github.com/stepanstipl/gpii-pure).

>> Very narrow scope - for example using --filter bindings.members:user: will result in only removing individual user permission

> Which is the goal of this task – to remove permissions added by :grant_org_admin. 

The task is called `set_org_perms` -> I would expect it to properly set org permissions. 

> Also, previous version of this code was only raising an exception when some permissions did not match with expected state of things, so it is already a progress :)

This approach is arguably less secure -> if we have extra permissions, let's say granting admin to `allUsers`, failing would show us there's some problem.

natarajaya commented:
> No, we would also use terraform for this

Looks like we have a consensus that this idea looks like an improvement, but now is not a good time to move `xk_infra` code to Terraform.

> This approach is arguably less secure -> if we have extra permissions, let's say granting admin to allUsers, failing would show us there's some problem

No, it is not – previous implementation [only cared about](https://github.com/gpii-ops/gpii-infra/blob/master/shared/rakefiles/xk_infra.rake#L114-L117) common SA and cloud-admin permissions.

natarajaya commented:
Since we already had a discussion about all pros and cons of this approach and our long-term plans for managing org-level and project-level IAM policies, I am going to merge this as is.
